zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/DedicatedZkClient.java

```diff
@@ -0,0 +1,465 @@
+    checkIfPathContainsShardingKey(path);
+    _rawZkClient.asyncCreate(path, datat, mode, cb);
+  }
+
+  @Override
+  public void asyncSetData(String path, Object datat, int version,
+      ZkAsyncCallbacks.SetDataCallbackHandler cb) {
+    checkIfPathContainsShardingKey(path);
+    _rawZkClient.asyncSetData(path, datat, version, cb);
+  }
+
+  @Override
+  public void asyncGetData(String path, ZkAsyncCallbacks.GetDataCallbackHandler cb) {
+    checkIfPathContainsShardingKey(path);
+    _rawZkClient.asyncGetData(path, cb);
+  }
+
+  @Override
+  public void asyncExists(String path, ZkAsyncCallbacks.ExistsCallbackHandler cb) {
+    checkIfPathContainsShardingKey(path);
+    _rawZkClient.asyncExists(path, cb);
+  }
+
+  @Override
+  public void asyncDelete(String path, ZkAsyncCallbacks.DeleteCallbackHandler cb) {
+    checkIfPathContainsShardingKey(path);
+    _rawZkClient.asyncDelete(path, cb);
+  }
+
+  @Override
+  public void watchForData(String path) {
+    checkIfPathContainsShardingKey(path);
+    _rawZkClient.watchForData(path);
+  }
+
+  @Override
+  public List<String> watchForChilds(String path) {
+    checkIfPathContainsShardingKey(path);
+    return _rawZkClient.watchForChilds(path);
+  }
+
+  @Override
+  public long getCreationTime(String path) {
+    checkIfPathContainsShardingKey(path);
+    return _rawZkClient.getCreationTime(path);
+  }
+
+  @Override
+  public List<OpResult> multi(Iterable<Op> ops) {
+    return _rawZkClient.multi(ops);
+  }
+
+  @Override
+  public boolean waitUntilConnected(long time, TimeUnit timeUnit) {
+    return _rawZkClient.waitUntilConnected(time, timeUnit);
+  }
+
+  @Override
+  public String getServers() {
+    return _rawZkClient.getServers();
+  }
+
+  @Override
+  public long getSessionId() {
+    return _rawZkClient.getSessionId();
+  }
+
+  @Override
+  public void close() {
+    _rawZkClient.close();
+  }
+
+  @Override
+  public boolean isClosed() {
+    return _rawZkClient.isClosed();
+  }
+
+  @Override
+  public byte[] serialize(Object data, String path) {
+    checkIfPathContainsShardingKey(path);
+    return _rawZkClient.serialize(data, path);
+  }
+
+  @Override
+  public <T> T deserialize(byte[] data, String path) {
+    checkIfPathContainsShardingKey(path);
+    return _rawZkClient.deserialize(data, path);
+  }
+
+  @Override
+  public void setZkSerializer(ZkSerializer zkSerializer) {
+    _rawZkClient.setZkSerializer(zkSerializer);
+  }
+
+  @Override
+  public void setZkSerializer(PathBasedZkSerializer zkSerializer) {
+    _rawZkClient.setZkSerializer(zkSerializer);
+  }
+
+  @Override
+  public PathBasedZkSerializer getZkSerializer() {
+    return _rawZkClient.getZkSerializer();
+  }
+
+  /**
+   * Checks whether the given path belongs matches the ZK path sharding key this DedicatedZkClient is designated to at initialization.
+   * @param path
+   * @return
+   */
+  private void checkIfPathContainsShardingKey(String path) {
+    if (!path.startsWith(_zkRealmShardingKey)) {
```

Comment:
If sharding key is `/abcde`, are these paths expected to pass this check and supposed to use this zkclient:
- `/abcde_0/xxx`
- `/abcde/abcde-0/xxx`

Reply:
Good catch! Let me fix this up.

Reply:
There are other possible way to fix it and reason it would work.

For example the prefix must be normalized to end with "/" before the match against input. 

Either way, this implementation detail can be fix later.

Reply:
We don't have all the components implemented to make this work yet. 

@kaisun2000 

> For example the prefix must be normalized to end with "/" before the match against input.

This may not be necessary since we'll have the trie-like routing data already. That way, we don't implement another line of the check logic and we are able to re-use a tested component.

At any rate, I'll put a TODO and let's revisit this method when we have all the components ready.